### ISSUE TITLE
Failed to get correct device id inside on_fit_start() or setup()

### ISSUE THREAD

**Issue by pableeto** [PyTorchLightning/pytorch-lightning#4849]
## 🐛 Bug
In DDP training mode, calling self.device in on_fit_start() or setup() function will return "cuda", while calling self.device in training_step will return "cuda:id" where id corrspondes to the correct device id. 

### To Reproduce
~~~
BoringModel(pl.LightningModule):
    def on_fit_start(self):
        print(self.device)      # will always return "cuda" regardless which device it on
    def setup(self, stage):
        print(self.device)      # will always return "cuda" regardless which device it on
    def training_step(self, **kwargs):
        print(self.device)      # will return device ids, e.g. "cuda:0"
~~~

### Expected behavior
on_fit_start() and/or setup() should be able to get correct device ids as same in training_step().

**Comment by justusschock** [comment]
Actually I think this is kind of the correct behaviour. `self.device` only reflects changes of device types from model parameters. 

If you call `model.cuda()` on your model, `self.device` only returns `cuda` (which means, the device should be the torch default gpu) and when you call `model.cuda(1)` your device will explicitly be the first gpu.

```
>>> model.device
device(type='cpu')
>>> model.cuda()
BoringModel()
>>> model.device
device(type='cuda')
>>> model.cuda(1)
BoringModel()
>>> model.device
device(type='cuda', index=1)
```

However, I opened a PR (#4851) to always include a device index there.
